File: python/221-230/Maximal Square.py
class Solution:
    # @param {character[][]} matrix
    # @return {integer}
    def maximalSquare(self, matrix):
        if len(matrix) == 0:
            return 0
        h = [0] * (len(matrix[0])+1)
        res = 0
        
        for i in range(len(matrix)):
            matrix[i].append('0')
            mystack = []
            
            for j in range(len(matrix[i])):
                if matrix[i][j] == '1':
                    h[j] += 1
                else:
                    h[j] = 0
                
                if len(mystack) == 0 or h[j] >= h[mystack[-1]]:
                    mystack.append(j)
                else:
                    while len(mystack) > 0 and h[j] < h[mystack[-1]]:
                        height = h[mystack[-1]]
                        del mystack[-1]
                        width = j if len(mystack) == 0 else j - 1 - mystack[-1]
                        side = min(height, width)
                        res = max(res, side*side)
                    mystack.append(j)
        
        return res

File: cpp/221-230/Maximal Square.cpp
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        if (matrix.empty() || matrix.front().empty()) return 0;
        int m = matrix.size();
        int n = matrix.front().size();
        vector<vector<int>> dp(m, vector<int>(n, 0));
        int maxLen = 0;
        
        for (int i = 0; i < m; i++) {
            for (int j = 0;j < n; j++) {
                if (matrix[i][j] == '1') {
                    dp[i][j] = 1;
                    if (i > 0 && j > 0) {
                        int tmp = min(dp[i-1][j-1], min(dp[i][j-1], dp[i-1][j]));
                        dp[i][j] = max(dp[i][j], tmp+1);
                    }
                }
                maxLen = max(maxLen, dp[i][j]);
            }
        }
        
        return maxLen*maxLen;
    }
};
